src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb

```diff
@@ -1647,17 +1647,18 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
             End SyncLock
 
             If completedCompilationUnit Then
-                EventQueue.Enqueue(New CompilationUnitCompletedEvent(Me, tree))
+                EventQueue.TryEnqueue(New CompilationUnitCompletedEvent(Me, tree))
             End If
 
             If completedCompilation Then
-                EventQueue.Enqueue(New CompilationCompletedEvent(Me))
-                EventQueue.Complete() ' signal the End Of compilation events
+                EventQueue.TryEnqueue(New CompilationCompletedEvent(Me))
+                EventQueue.PromiseNotToEnqueue()
+                EventQueue.TryComplete() ' signal the End Of compilation events
             End If
         End Sub
 
         Friend Function ShouldAddEvent(symbol As Symbol) As Boolean
-            If EventQueue Is Nothing Then
+            If EventQueue Is Nothing OrElse EventQueue.IsCompleted Then
```

Comment:
The two conditions "promised not to enqueue" and "cancelled/completed" are separate and have nothing to do with each other.

"promise not to enqueue" occurs when the compiler is in a state where all events that could possibly be produced should have already been produced. It is checked in the queue, and it causes an exception to be raised if there is an attempt to enqueue an event.

Cancelled, or Completed, can occur because the client says they are not interested in any further events. In this case we just stop producing events due to the client's declaration. That is what this new check is for.